src/core/eme/index.ts

```diff
@@ -143,45 +136,79 @@ function createEME(
       type: "string",
     }));
   }
 
-  // get the MediaKeys element the associated options
-  // This is done as the "encrypted" event is not yet received for performance
-  const mediaKeysInfos$ = findCompatibleKeySystem(keySystems, instanceInfos)
-    .mergeMap((keySystemAccessInfos) => {
-      return createMediaKeysObs(keySystemAccessInfos.keySystemAccess)
-        .mergeMap(function prepareMediaKeysConfiguration(mediaKeys) {
-          const { keySystem } = keySystemAccessInfos;
-          const { serverCertificate } = keySystem;
-          setSessionStorage(keySystem);
+  /**
+   * Encrypted event may be distinct until initData changes.
+   * It is usefull to trigger session creation on two consecutive init datas.
+   */
+  const encrypted$ = onEncrypted$(video)
+    .distinctUntilChanged(function isSameEncryptedEvent(prevEvt, evt) {
+      if (
+        evt.initData == null || prevEvt.initData == null ||
+        evt.initDataType !== prevEvt.initDataType
+      ) {
+        return false;
+      }
+      const evtHash = hashInitData(new Uint8Array(evt.initData));
+      const prevHash = hashInitData(new Uint8Array(prevEvt.initData));
+      return evtHash === prevHash;
+    });
 
-          const _mediaKeysInfos$ = Observable
-            .of(objectAssign({ mediaKeys }, keySystemAccessInfos));
+  // Create or get cached session.
+  return  Observable.combineLatest(
+    encrypted$,
+    createMediaKeys(keySystems, errorStream)
+  ).concatMap(
+    ([encryptedEvent, mediaKeysInfos]) => {
+      log.info("eme: encrypted event", encryptedEvent);
 
-          if (
-            serverCertificate != null &&
-            typeof mediaKeys.setServerCertificate === "function"
-          ) {
-            return trySettingServerCertificate(mediaKeys, serverCertificate, errorStream)
-              .concat(_mediaKeysInfos$);
-          }
-          return _mediaKeysInfos$;
-        });
-    });
+      if (encryptedEvent.initData == null) {
+        const error = new Error("no init data found on media encrypted event.");
+        throw new EncryptedMediaError("INVALID_ENCRYPTED_EVENT", error, true);
+      }
 
-  return Observable.combineLatest(
-    onEncrypted$(video).take(1),
-    mediaKeysInfos$
-  )
-    .mergeMap(([encryptedEvent, mediaKeysInfos]) =>
-      handleEncryptedEvent(encryptedEvent, mediaKeysInfos, video, errorStream)
-    );
+      const events$ = createOrReuseSessionWithRetry(
+        encryptedEvent,
+        mediaKeysInfos
+      );
+
+      return Observable.combineLatest(
+        events$,
+        Observable.of(encryptedEvent),
+        Observable.of(mediaKeysInfos)
+      );
+    }
+  // Handle events from created session
+  ).mergeMap((
+      [event, encryptedEvent, mediaKeysInfos]:
+        [ISessionEvent, MediaEncryptedEvent, IMediaKeysInfos],
+      i: number
+    ) => {
+      const setMediaKeys$ = i === 0 ?
+        setMediaKeysObs(mediaKeysInfos, video, instanceInfos) :
+        Observable.empty();
```

Comment:
So from there, we only set the MediaKeys to the video element once at least one Event was set by the session.

From how I can see it from here, this could lead to a dead lock in some browser if we change the list of session events emitted.